### ISSUE TITLE
[Request] Please make ytfzf search again after playback ends

### ISSUE THREAD

**Issue by pitsi** [pystardust/ytfzf#244]
Debian testing x64, python 3.9.2, ytfzf 1.1.5 and all settings set to default in conf.sh except those 3
```
enable_hist=0
enable_loop=1
video_pref="140"
```
The first is because I do not see anything useful in keeping a history of what was played and the third one because I prefer to only listen to audio, and "140" refers to the "m4a audio only" stream of each video. The second one is explained below.

So, when enable_loop is set to 0, the app exits once playback ends and I have to start it again, search again etc. When enable_loop is set to 1, the app returns to the search results and all I can do is to select something from the existing results. If I want to search for something new, I have to either exit the app with ctrl+c and start over.

Can you please make ytfzf, via a config parameter I assume, search again after the playback ends?

**Comment by Euro20179** [comment]
I think you are looking for `search_again`

**Comment by pitsi** [comment]
Wow thanks! Can yo please make it a setting in conf.sh so as to make it permanent?

**Comment by pystardust** [comment]
Yes, you could add this to your config file `conf.sh`
```
search_again=1
```

**Comment by pitsi** [comment]
It worked! Thanks a lot. 
Can I add any command line parameter in conf.sh like so?

**Comment by Euro20179** [comment]
> Can I add any command line parameter in conf.sh like so?

all command line parameters should have some way to be set in conf.sh,
check `/docs/conf.sh` to get a list of every variable

**Comment by pitsi** [comment]
It is not in the current conf.sh, that's why I asked. And that is also the reason I opened this request, because I thought all the available parameters/options are included in conf.sh and none seemed to do what I wanted.

**Comment by pitsi** [comment]
So, I just noticed that ytfzf was updated to v2.0 today. I now have all these in my ~/.config/ytfzf/conf.sh
```
$ cat .config/ytfzf/conf.sh 
enable_hist=0
#enable_loop=1
video_pref=140
search_again=1
```
I have already checked ytfzf.5 for all the paremeters and the `search_again` parameter seems to have been removed. Was it replaced by another one? The `enable_loop` one was probably changed to `is_loop`, right?
Also, the very convenient `-q` parameter has also been removed. Is there something that replaces it?
```
$ ytfzf -q
Illegal option -q
Search
> 
```

**Comment by Euro20179** [comment]
> Also, the very convenient -q parameter has also been removed. Is there something that replaces it?

You can add this to your config, instead of `-q` it'll be `--search-hist` (you could replace `search-hist` in the `on_opt_parse` to q but you'll have to use `--q`

```sh
#if multiple scrapers are enabled, you'll have the search being saved multiple times
has_search=0
on_search () {
    [ $has_searched -eq 0 ] && {
        has_searched=1
        echo "$search" >> "$file_you_want_to_save_as_search_history"
    }
}
on_opt_parse (){
    case "$1" in
        search-hist) 
            search="$(cat "$file_you_want_to_save_as_search_history" | fzf)"
            return 1
            ;;
    esac
}
```

I'll look into adding this back into the script, depending on how much complexity it adds i'll do it.

> The enable_loop one was probably changed to is_loop, right?

Yes

> I have already checked ytfzf.5 for all the paremeters and the search_again parameter seems to have been removed.

This was removed cause all it really did was open a prompt menu after you close fzf, but it added a lot of complexity
if you want to replicate it you could make a wrapper script that has this in it:
```sh
while true; do
    ytfzf
done
```
(running ytfzf with no search still brings up a search prompt)

**Comment by Euro20179** [comment]
> I'll look into adding this back into the script, depending on how much complexity it adds i'll do it.

It seems to add a little complexity in the function where history is cleared, and the --history-clear flag.
But it's not enough to say no to this.

#438, test it out and tell me how it is.
download:
```sh
git clone https://github.com/euro20179/ytfzf
cd ytfzf
git checkout search-hist

./ytfzf ..
```

**Comment by pitsi** [comment]
> You can add this to your config, instead of `-q` it'll be `--search-hist` (you could replace `search-hist` in the `on_opt_parse` to q but you'll have to use `--q`

I don't get this, sorry. Launching ytfzf with those parameters results in the similar errors. What am I doing wrong?
```
$ ytfzf --q
/home/x/bin/ytfzf: illegal long option -- q
Search
> ^C
$ ytfzf --search-hist
/home/x/bin/ytfzf: illegal long option -- search-hist
Search
> ^C
```
I will change the other parameters as suggested.

**Comment by Euro20179** [comment]
> Or at least choose a different invidious instance if I want

Yes (`invidious_instance`), make sure to include the `https://` part in the link, and I believe I made it flexible enough to where you can host your own invidious instance and put `invidious_instance=http://localhost`

> I have some bad memories from instances that were working great one day and not at all the next one.

same

> Can I switch it to youtube

Youtube is basically just worse invidious, and it makes it harder to implement basically everything.
Although with subscriptions there is both -cSI, and -cS (SI = invidious, S = youtube), because they both have advantages and disadvantages. (Invidious sometimes hangs, youtube sometimes ends up treating me like a bot, idk if it happens to other people)

However I did write a scraper for youtube to see what would happen, you can download [this file](https://gist.github.com/Euro20179/4f91fe0cc3b8ff00b1afdf7d429339ae) and put it in ~/.config/ytfzf/scrapers and then run `ytfzf -cyt <search>`

> I don't get this, sorry. Launching ytfzf with those parameters results in the similar errors. What am I doing wrong?

Did you add the code snippet to your config?

**Comment by pitsi** [comment]
Happy new year everyone and sorry for the huge delay in replying.
The youtube scrapper does work, although sometimes an 500 http error like this one comes up. Yt-dlp is installed obviously.
```
Scraping Youtube...
[ytdl_hook] ERROR: [generic] Unable to download webpage: HTTP Error 500: Internal Server Error (caused by <HTTPError 500: 'Internal Server Error'>); please report this issue on  https://github.com/yt-dlp/yt-dlp . Make sure you are using the latest version; see  https://github.com/yt-dlp/yt-dlp  on how to update. Be sure to call yt-dlp with the --verbose flag and include its complete output. (caused by <HTTPError 500: 'Internal Server Error'>); please report this issue on  https://github.com/yt-dlp/yt-dlp . Make sure you are using the latest version; see  https://github.com/yt-dlp/yt-dlp  on how to update. Be sure to call yt-dlp with the --verbose flag and include its complete output. 
[ytdl_hook] youtube-dl failed: unexpected error occurred 
Failed to recognize file format.
```
It would be nice if ytfzf could use the scrapper by default. In case there api keys that are needed to do so, I have made mien for kodi and I have no problem using them on a different project that must access that (garbage) api of youtube.

Similarly, the invidious instance I used (which sadly is not yewtu.be because it blocks access to its api) sometimes pops this error, but it plays the file at least.
```
Search
> anser
Scraping YouTube (with https://some-lesser-known-instance) (anser, pg: 1)
[ffmpeg] https: HTTP error 500 Internal Server Error
Failed to open https://invidious.domain.glass/watch?v=C75GfC1y_nw.
 (+) Audio --aid=1 --alang=eng (*) (aac 2ch 44100Hz)
AO: [alsa] 48000Hz stereo 2ch float
A: 00:00:42 / 00:03:33 (20%) Cache: 169s/6MB
```
As for the code snippet that will enable history in ytfzf, I am sorry but my brain cells are not enough to do the needed work :(
That is why I downloaded the newer version of ytfzf (commit c157149)  that contains the changes. It seems to work because 
```
$ ytfzf -q
You have no search history
```
but it would be nicer if it used the existing ~/.cache/ytfzf/ytfzf_search_hst (mine is like 500 lines big) from v1.x.

There is also this notification that comes up once playback is started. It looks cool, but it would be cooler to have the name of the video on it. Or an option to disable it, since I do not want that useless info. 
![2021-12-31-143325_1280x1024_scrot](https://user-images.githubusercontent.com/5880037/147884520-55c71157-80c9-412d-bc1d-3f1a200c80c3.png)

Last but not least, after a few days of usage, I think that `is_loop=1` is not as convenient as `search_again=1` was and the reason is simple. The first parameter shows the results of the search again, so if I want to search for something different, I have to exit ytfzf and start it again.

p.s. I may revert to v1.x in the following days and keep v2 as ytfzf2 for the testing that has to be done.

**Comment by Euro20179** [comment]
Happy 2022!

> It would be nice if ytfzf could use the scrapper by default

Problem with the youtube scraper is it's just harder to support filters, and --pages isn't possible.

> Similarly, the invidious instance I used (which sadly is not yewtu.be because it blocks access to its api) sometimes pops this error, but it plays the file at least.

This error and the other one are the same error I think, not sure what's causing it.

> but it would be nicer if it used the existing ~/.cache/ytfzf/ytfzf_search_hst from v1.x. Mine is like 500 lines big)

The old format and new format are different, I can try writing a converter script and posting it here.

> There is also this notification that comes up once playback is started. It looks cool, but it would be cooler to have the name of the video on it. Or an option to disable it, since I do not want that useless info.

It is possible to disable it, but it's a bit of effort, I may refactor it to make it easier. Because disabling it currently requires rewriting the video_player, audio_player, and downloader functions in your config file. (in total it's like 20 lines of code)

> Last but not least, after a few days of usage, I think that is_loop=1 is not as convenient as search_again=1

I'll look into adding this, and if it's not too complex i'll merge it.


Thank you for all the suggestions.

**Comment by Euro20179** [comment]
> I'll look into adding this, and if it's not too complex i'll merge it.

I looked into adding it, I'm not sure if it'll be complex, but it will require a decently sized refactor, in order to add it cleanly.

**Comment by Euro20179** [comment]
Aight, I've got search_again working. #443 I'd highly recommend trying it to make sure it works
```sh
git clone https://github.com/euro20179/ytfzf
cd ytfzf
git checkout organize-code
chmod +x ytfzf
./ytfzf -s ....
```

**Comment by Euro20179** [comment]
I recently made a commit, where `-l` now affects `-s`, `-s` alone will search again after quitting the video, `-sl` will search again only after leaving `fzf` (or whichever menu your using)

I believe this was not the case in `1.x`

**Comment by pitsi** [comment]
So, I upgraded to 2.1-rc3 (as of today, I first updated it at rc1). The `-q` parameter works although it uses a new history file. At first, I thought about copying the old file with the new name, but their entries differ a lot. On the old file it looks like so
`2022-01-16 08:46:10	anser`
and on the new one it has no timestamp
`anser`
Moreover, it returns to search when the playback is done, which is quite convenient.

The notification no longer comes up, which is nice. However, it also does not come up even if I set `notify_video_playing` to 1. No idea why.

And, as last time, the invidious instance I used has its issues. Sometimes it does search, sometimes it does not. Sometimes it plays, sometimes it does not. But this is a "feature" with any invidious instance.

I hope I did not miss anything from the above.

**Comment by Euro20179** [comment]
>> On the old file it looks like so
`2022-01-16 08:46:10 anser`
and on the new one it has no timestamp

To be honest, I forgot the old implementation had time stamps. Since I haven't released 2.1 yet I'll modify what it writes, and  I'll try to make it so you can customize what it writes.

> The notification no longer comes up, which is nice. However, it also does not come up even if I set `notify_video_playing` to 1. No idea why.

The variable is `notify_playing`, I messed up the man page sorry.

> And, as last time, the invidious instance I used has its issues. Sometimes it does search, sometimes it does not. Sometimes it plays, sometimes it does not. But this is a "feature" with any invidious instance.

I've written a yt scraper, mainly to keep the code as a backup, I can share it here, and instructions on how to use it in a few hours, but it's missing all features that the invidious scrape has.

**Comment by pitsi** [comment]
You mentioned the yt scraper above. I still have it in ~/.config/ytfzf/scrapers. Has anything changed in it? Can I make ytfzf use it every time without that `-cyt` parameter? I assume it is something like
`scrape=youtube`

The notification thing works as it should now. Thanks for the fix.

**Comment by Euro20179** [comment]
> You mentioned the yt scraper above. I still have it in ~/.config/ytfzf/scrapers

Oh I knew I gave it to someone but I forgot who. Nothing has changed.

> I assume it is something like
> `scrape=youtube`

The name of the scrape should match the name of the file, so `scrape=yt`

`youtube` or `Y` is reserved for invidious.

**Comment by Euro20179** [comment]
> To be honest, I forgot the old implementation had time stamps. Since I haven't released 2.1 yet I'll modify what it writes, and I'll try to make it so you can customize what it writes.

5f795d6f8f7776a89ab85db702f051b4b1854998,

you can update to rc-4

**Comment by pitsi** [comment]
I upgraded to 2.1rc4. The yt scraper works with the suggested entry, although it is slow, but who cares :P
I will check the history "transplant" tomorrow.

**Comment by pitsi** [comment]
Entry in the old history file
`2021-11-02 08:16:33	vevilos`
vs similar entry in the new history file
`01/19/22 19:33:41	vevilos`
Is this intentional? Will it mess things up if I do it? Btw, my plan was to just 
`cp ytfzf_search_hst search_hist`

**Comment by Euro20179** [comment]
> Will it mess things up if I do it?

No as long as the timestamp and search are separated by a tab.

> Btw, my plan was to just

I only suggested doing it the other way so that your newest searches saved in `search_hist` are saved, but you can do it however you want.

>  The yt scraper works with the suggested entry, although it is slow

does it take longer than using `-cyt`? Or is it just that the yt scraper itself is slow?

**Comment by pitsi** [comment]
I managed to successfully transfer the history file and now everything works as it should, so I am closing the issue.

**Comment by pitsi** [comment]
I just noticed that `ytfzf -q` no longer gets me back to history after playback ends. Is this the default behavior of 2.x? Because it was working like so on 1.x. This is my configuration
```
$ cat .config/ytfzf/conf.sh 
enable_hist=0
#search_again=1
#is_loop=1
video_pref=140
#video_pref=22
#invidious_instance=https://some-lesser-known-instance
notify_playing=1
scrape=yt
```
With the above, when playback ends, ytfzf exits back to prompt. If the `search_again=1` is enabled, then it goes back to search as expected. Is there some (new?) variable that will get me back to the huge list of `ytfzf -q`?

**Comment by Euro20179** [comment]
I made it so that when you use -q it only does it once, because I thought doing it for every search would add code complexity, but I implemented it just now, and it didn't so 68ef357b5cfd1e2721d1b154161ea4e5d6a47139, feel free to try out this version. (when you use -l and -s, itll first go back to the menu, then when you exit the menu it searches again)
